modules/core/src/main/java/org/apache/ignite/spi/ExponentialBackoffTimeoutStrategy.java

```diff
@@ -0,0 +1,158 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.spi;
+
+import java.util.Date;
+
+/**
+ * Strategy which incorporates retriable network operation, handling of totalTimeout logic.
+ * It increases startTimeout based on exponential backoff algorithm.
+ *
+ * If failure detection is enabled it relies on totalTimeout
+ * otherwise implements exponential backoff totalTimeout logic based on startTimeout, maxTimeout and retryCnt.
+ */
+public class ExponentialBackoffTimeoutStrategy implements TimeoutStrategy {
+    /** Initial startTimeout, ms. */
+    private final long startTimeout;
+
+    /** Max startTimeout of the next try, ms. */
+    private final long maxTimeout;
+
+    /** Max reconnections count. */
+    private final int reconCnt;
+
+    /** Total startTimeout, ms. */
+    private final long totalTimeout;
+
+    /** Start of operation to check totalTimeout. */
+    private final long start;
+
+    /** Current startTimeout, ms. */
+    private long currTimeout;
+
+    /** Backoff coeffient to calculate next timeout.*/
+    private double backoffCoeff;
+
+    /** Compute expected backoffConnTimeout delay based on startTimeout, maxTimeout and reconCnt and backoff coeffient. */
+    public static long maxBackoffTimeout(
+            long startTimeout,
+            long maxTimeout,
+            long reconCnt,
+            double backoffCoeff
+    ) {
+        long maxBackoffTimeout = startTimeout;
+
+        for (int i = 1; i < reconCnt && maxBackoffTimeout < maxTimeout; i++)
+            maxBackoffTimeout += nextTimeout(maxBackoffTimeout, maxTimeout, backoffCoeff);
+
+        return maxBackoffTimeout;
+    }
+
+    /**
+     *
+     * @param timeout Timeout.
+     * @param maxTimeout Maximum startTimeout for backoff function.
+     * @param backoffCoeff Coefficient to calculate next backoff startTimeout.
+     * @return Next exponetial backoff totalTimeout.
+     */
+    public static long nextTimeout(long timeout, long maxTimeout, double backoffCoeff) {
+        return (long) Math.min(timeout * backoffCoeff, maxTimeout);
+    }
+
+    /**
+     *
+     * @param totalTimeout Total startTimeout.
+     * @param startTimeout Initial connection totalTimeout, will be taken for initial connection and handshake timeouts.
+     * @param maxTimeout Max connection Timeout.
+     * @param reconCnt Max number of reconnects.
+     */
+    public ExponentialBackoffTimeoutStrategy(
+        long totalTimeout,
+        long startTimeout,
+        long maxTimeout,
+        int reconCnt,
+        double backoffCoeff
+    ) {
+        this.totalTimeout = totalTimeout;
+
+        this.startTimeout = startTimeout;
+        this.maxTimeout = maxTimeout;
+
+        this.backoffCoeff = backoffCoeff;
+
+        currTimeout = startTimeout;
+
+        this.reconCnt = reconCnt;
+
+        start = System.currentTimeMillis();
```

Comment:
U.currentTimeMillis()

Reply:
I've changed to System.currentTimeMillis().

Reply:
This is intentional.

Reply:
Fixed